### PR TITLE
[SPARK-24257][SQL]LongToUnsafeRowMap calculate the new size may be wrong

### PR DESCRIPTION
**Pull request by cxzl25**
## What changes were proposed in this pull request?

LongToUnsafeRowMap
Calculate the new size simply by multiplying by 2
At this time, the size of the application may not be enough to store data
Some data is lost and the data read out is dirty

## How was this patch tested?
HashedRelationSuite
test("LongToUnsafeRowMap with big values")

### CODE REVIEW

**Comment by maropu** [comment]
@gatorsmile @hvanhovell Could you trigger tests?

**Comment by kiszk** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 1]
```diff
@@ -568,13 +568,16 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
...
+    if (needSize > nowSize) {
       val used = page.length
       if (used >= (1 << 30)) {
         sys.error("Can not build a HashedRelation that is larger than 8G")
       }
-      ensureAcquireMemory(used * 8L * 2)
-      val newPage = new Array[Long](used * 2)
+      val multiples = math.max(math.ceil(needSize.toDouble / (used * 8L)).toInt, 2)
+      ensureAcquireMemory(used * 8L * multiples)
+      val newPage = new Array[Long](used * multiples)
```
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Do we move the size check into before the allocation? IIUC, we have to check `used * multiplies` <= `ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH` now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 2]
```diff
@@ -568,13 +568,16 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
...
+    val nowSize = page.length * 8L + Platform.LONG_ARRAY_OFFSET
+    if (needSize > nowSize) {
       val used = page.length
       if (used >= (1 << 30)) {
         sys.error("Can not build a HashedRelation that is larger than 8G")
       }
-      ensureAcquireMemory(used * 8L * 2)
-      val newPage = new Array[Long](used * 2)
+      val multiples = math.max(math.ceil(needSize.toDouble / (used * 8L)).toInt, 2)
+      ensureAcquireMemory(used * 8L * multiples)
```
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Do we move the size check into before the allocation? IIUC, we have to check `used * multiplies` <= ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH` now.

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: How about shaping up this logic along with the other similar ones (spliting this func into two parts: `grow`/`append`)? e.g., `UTF8StringBuilder`  https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UTF8StringBuilder.java#L43

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 3]
```diff
@@ -568,13 +568,16 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
     }
 
     // There is 8 bytes for the pointer to next value
-    if (cursor + 8 + row.getSizeInBytes > page.length * 8L + Platform.LONG_ARRAY_OFFSET) {
+    val needSize = cursor + 8 + row.getSizeInBytes
+    val nowSize = page.length * 8L + Platform.LONG_ARRAY_OFFSET
+    if (needSize > nowSize) {
       val used = page.length
       if (used >= (1 << 30)) {
         sys.error("Can not build a HashedRelation that is larger than 8G")
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: This is not related to this pr though, `sys.error` instead of `UnsupportedOperationException`?
https://github.com/apache/spark/blob/b6c50d7820aafab172835633fb0b35899e93146b/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UTF8StringBuilder.java#L45

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: +1 on grow/append

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 4]
```diff
@@ -568,13 +568,16 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
...
       }
-      ensureAcquireMemory(used * 8L * 2)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Doubling the size when growing is very typical, seems what you want to address is when the memory is enough for the requsted size but not enough for doubling the size. I'd suggest we should double the size most of the time, as long as there is enough memory.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 3 above)
Comment (cxzl25) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: ok. sys.error instead of UnsupportedOperationException

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 4 above)
Comment (cxzl25) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: ok . Doubling the size when growing.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala (hunk 2 above)
Comment (cxzl25) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: ok.Spliting append func into two parts: grow/append.

**Comment by cxzl25** [comment]
Thanks for your review. @maropu @kiszk @cloud-fan 

I submitted a modification including the following:
1. spliting append func into two parts:grow/appendG
2. doubling the size when growing
3. sys.error instead of UnsupportedOperationException

**Comment by JoshRosen** [comment]
@cxzl25, to clarify:

> Some data is lost and the data read out is dirty

To clarify, is this a potential cause of a wrong-answer correctness bug? If so, we should be sure to backport the resulting fix to maintenance branches. /cc @cloud-fan @gatorsmile

**Comment by cxzl25** [comment]
@JoshRosen  @cloud-fan @gatorsmile 
When introducing [SPARK-10399](https://issues.apache.org/jira/browse/SPARK-10399),UnsafeRow#getUTF8String check the size at this time.
[UnsafeRow#getUTF8String](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java#L420)
[OnHeapMemoryBlock](https://github.com/apache/spark/blob/master/common/unsafe/src/main/java/org/apache/spark/unsafe/memory/OnHeapMemoryBlock.java#L34)
>The sum of size 2097152 and offset 32 should not be larger than the size of the given memory space 2097168
image

But when this patch is not introduced, no error, get wrong value.

spark-2 2 0

**Comment by cloud-fan** [comment]
> Calculate the new size simply by multiplying by 2
At this time, the size of the application may not be enough to store data
Some data is lost and the data read out is dirty

Can you explain more about it? IIUC if we don't have enough memory for `size * 2`, we would just fail with OOM, instead of setting a wrong size.

**Comment by cxzl25** [comment]
@cloud-fan 
LongToUnsafeRowMap#append(key: Long, row: UnsafeRow)
when row.getSizeInBytes > newPageSize( oldPage.length * 8L * 2),still use newPageSize value.
When the new page size is insufficient to hold the entire row of data, Platform.copyMemory is still called.No error.
At this time, the last remaining data was discarded.
When reading data, read this buffer according to offset and length,the last data is unpredictable.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 5]
```diff
@@ -626,6 +618,32 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
     }
   }
 
+  private def grow(neededSize: Int): Unit = {
+    // There is 8 bytes for the pointer to next value
+    val totalNeededSize = cursor + 8 + neededSize
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: nit:
```
private def grow(inputRowSize: Int): Unit = {
  val neededSize = cursor + 8 + inputRowSize
  val currentSize = page.length * 8L + Platform.LONG_ARRAY_OFFSET
  ...
}
```

Comment (cxzl25) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: @cloud-fan Thank you for your suggestion and code.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 6]
```diff
@@ -626,6 +618,29 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
     }
   }
 
+  private def grow(inputRowSize: Int): Unit = {
+    val neededNumWords = (cursor - Platform.LONG_ARRAY_OFFSET + 8 + inputRowSize + 7) / 8
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: don't forget the comment for the 8 bytes pointer

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 7]
```diff
@@ -626,6 +618,29 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
...
+    if (neededNumWords > page.length) {
+      if (neededNumWords > (1 << 30)) {
+        throw new UnsupportedOperationException(
+          "Can not build a HashedRelation that is larger than 8G")
+      }
+      val newNumWords = math.max(neededNumWords, math.min(page.length * 2, 1 << 30))
+      if (newNumWords > ARRAY_MAX) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: we won't need this check now, `newNumWords` is guaranteed to be less than (1 << 30), which is much smaller than `ARRAY_MAX`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 8]
```diff
@@ -254,6 +254,30 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+      new StaticMemoryManager(
+        new SparkConf().set(MEMORY_OFFHEAP_ENABLED.key, "false"),
+        Long.MaxValue,
+        Long.MaxValue,
+        1),
+      0)
+    val unsafeProj = UnsafeProjection.create(Seq(BoundReference(0, StringType, false)))
+    val keys = Seq(0L)
+    val map = new LongToUnsafeRowMap(taskMemoryManager, 1)
+    val bigStr = UTF8String.fromString("x" * 1024 * 1024 * 2)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: let's add a comment to say, the page array is initialized with length `1 << 17`, so here we need a value larger than `1 << 18`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 9]
```diff
@@ -254,6 +254,30 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+    keys.foreach { k =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: we just have one key, why use loop?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 10]
```diff
@@ -30,6 +30,7 @@ import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.plans.physical.BroadcastMode
 import org.apache.spark.sql.types.LongType
 import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.array.ByteArrayMethods
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: not needed

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 11]
```diff
@@ -254,6 +254,30 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+      map.append(k, unsafeProj(InternalRow(bigStr)))
+    }
+    map.optimize()
+    val row = unsafeProj(InternalRow(bigStr)).copy()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: `val resultRow = new UnsafeRow(1)`

**Comment by cloud-fan** [comment]
LGTM, good catch!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala [hunk 12]
```diff
@@ -254,6 +254,30 @@ class HashedRelationSuite extends SparkFunSuite with SharedSQLContext {
...
+        Long.MaxValue,
+        1),
+      0)
+    val unsafeProj = UnsafeProjection.create(Array[DataType](StringType))
+    val map = new LongToUnsafeRowMap(taskMemoryManager, 1)
+
+    val key = 0L
+    // the page array is initialized with length 1 << 17 (1M bytes),
+    // so here we need a value larger than 1 << 18 (2M bytes),to trigger the bug
+    val bigStr = UTF8String.fromString("x" * (1 << 22))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: to double check, do we have to use `1 << 22` to trigger this bug?

Comment (cxzl25) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: Not necessary. 
Just chose a larger value to make it easier to lose data.

**Comment by maropu** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala (hunk 12 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: do you mean this bug can't be reproduced consistently? e.g. if we pick `1 << 18 + 1`, we may not expose this bug, so we have to use `1 << 22` to  100% reproduce this bug?

**Comment by kiszk** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala (hunk 12 above)
Comment (cxzl25) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: LongToUnsafeRowMap#getRow
resultRow=UnsafeRow#pointTo(page(1<<18), baseOffset(16), sizeInBytes(1<<21+16))

UTF8String#getBytes
copyMemory(base(page), offset, bytes, BYTE_ARRAY_OFFSET, numBytes(1<<21+16));

In the case of similar size sometimes, can still read the original value.

When introducing SPARK-10399,UnsafeRow#getUTF8String check the size at this time.
If we pick 1 << 18 + 1, 100% reproduce this bug.

But when this patch is not introduced, differences that are too small sometimes do not trigger.
So I chose a larger value.

My understanding may be problematic. Please advise. Thank you.

```java
        sun.misc.Unsafe unsafe;
        try {
            Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
            unsafeField.setAccessible(true);
            unsafe = (sun.misc.Unsafe) unsafeField.get(null);
        } catch (Throwable cause) {
            unsafe = null;
        }

        String value = "xxxxx";
        byte[] src = value.getBytes();

        byte[] dst = new byte[3];
        byte[] newDst = new byte[5];

        unsafe.copyMemory(src, 16, dst, 16, src.length);
        unsafe.copyMemory(dst, 16, newDst, 16, src.length);

        System.out.println("dst:" + new String(dst));
        System.out.println("newDst:" + new String(newDst));
```
output:
>dst:xxx
>newDst:xxxxx

**Comment by kiszk** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala (hunk 12 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/HashedRelationSuite.scala]: then 1 << 19 should be good enough as it doubles the size?

**Comment by vanzin** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3/2.2/2.1/2.0! There is no conflict so I backported all the way to 2.0. I'll watch the jenkins build in the few days.

**Comment by cxzl25** [comment]
@cloud-fan Thank you very much for your help.
